tortoise/verifying.go

```diff
@@ -0,0 +1,143 @@
+package tortoise
+
+import (
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+func newVerifying(config Config, common *commonState) *verifying {
+	return &verifying{
+		Config:       config,
+		commonState:  common,
+		goodBallots:  map[types.BallotID]struct{}{},
+		layerWeights: map[types.LayerID]weight{},
+		totalWeight:  weightFromUint64(0),
+	}
+}
+
+type verifying struct {
+	Config
+	*commonState
+
+	goodBallots map[types.BallotID]struct{}
+	// weight of good ballots in the layer
+	layerWeights map[types.LayerID]weight
+	// total weight of good ballots from verified + 1 up to last
+	totalWeight weight
+}
+
+func (v *verifying) processLayer(logger log.Log, lid types.LayerID, ballots []tortoiseBallot) {
+	counted, count := v.sumGoodBallots(logger, ballots)
+
+	// TODO(dshulyak) counted weight should be reduced by the uncounted weight per conversation with research
+
+	v.layerWeights[lid] = counted
+	v.totalWeight = v.totalWeight.add(counted)
+
+	logger.With().Info("computed weight from good ballots",
+		log.Stringer("weight", counted),
+		log.Stringer("total", v.totalWeight),
+		log.Stringer("expected", v.epochWeight[lid.GetEpoch()]),
+		log.Int("ballots_count", len(ballots)),
+		log.Int("good_ballots_count", count),
+	)
+}
+
+func (v *verifying) verifyLayers(logger log.Log) types.LayerID {
+	localThreshold := weightFromUint64(0)
+	// TODO(dshulyak) expected weight for local threshold should be based on last layer.
+	localThreshold = localThreshold.add(v.epochWeight[v.processed.GetEpoch()])
+	localThreshold = localThreshold.fraction(v.LocalThreshold)
+
+	for lid := v.verified.Add(1); lid.Before(v.processed); lid = lid.Add(1) {
+		// TODO(dshulyak) expected weight must be based on the last layer.
+		threshold := computeExpectedWeight(v.epochWeight, lid, v.processed)
+		threshold = threshold.fraction(v.GlobalThreshold)
+		threshold = threshold.add(localThreshold)
+
+		votingWeight := weightFromUint64(0)
+		votingWeight = votingWeight.add(v.totalWeight)
+		votingWeight = votingWeight.sub(v.layerWeights[lid])
+
+		layerLogger := logger.WithFields(
+			log.Stringer("candidate_layer", lid),
+			log.Stringer("voting_weight", votingWeight),
+			log.Stringer("threshold", threshold),
+			log.Stringer("local_threshold", localThreshold),
+		)
+
+		// 0 - there is not enough weight to cross threshold.
+		// 1 - layer is verified and contextual validity is according to our local opinion.
+		if votingWeight.cmp(threshold) == abstain {
+			layerLogger.With().Warning("candidate layer is not verified. voting weight is lower than the threshold")
+			return lid.Sub(1)
+		}
+
+		// if there is any block with neutral local opinion - we can't verify the layer
+		// if it happens outside of hdist - protocol will switch to full tortoise
+		for _, bid := range v.blocks[lid] {
```

Comment:
why is this check necessary when you have the check on line 70 `if votingWeight.cmp(threshold) == abstain {`

Reply:
on line 70 i check if i have enough weight to verify the layer, but the blocks in a layer can be undecided if the layer is still within zdist